tests/beta/tapes/test_rev.py

```diff
@@ -0,0 +1,601 @@
+
+        # analytic gradient
+        grad_fn = qml.grad(cost)
+        grad_A = grad_fn(reused_p, other_p)
+
+        # manual gradient
+        grad_true0 = (
+            expZ(
+                Rx(reused_p) @ Rz(other_p) @ Ry(reused_p + np.pi / 2) @ Rx(extra_param) @ zero_state
+            )
+            - expZ(
+                Rx(reused_p) @ Rz(other_p) @ Ry(reused_p - np.pi / 2) @ Rx(extra_param) @ zero_state
+            )
+        ) / 2
+        grad_true1 = (
+            expZ(
+                Rx(reused_p + np.pi / 2) @ Rz(other_p) @ Ry(reused_p) @ Rx(extra_param) @ zero_state
+            )
+            - expZ(
+                Rx(reused_p - np.pi / 2) @ Rz(other_p) @ Ry(reused_p) @ Rx(extra_param) @ zero_state
+            )
+        ) / 2
+        expected = grad_true0 + grad_true1  # product rule
+
+        assert np.allclose(grad_A[0], expected, atol=tol, rtol=0)
+
+    def test_gradient_gate_with_multiple_parameters(self, tol):
+        """Tests that gates with multiple free parameters yield correct gradients."""
+        x, y, z = [0.5, 0.3, -0.7]
+
+        with ReversibleTape() as tape:
+            qml.RX(0.4, wires=[0])
+            qml.Rot(x, y, z, wires=[0])
+            qml.RY(-0.2, wires=[0])
+            expval(qml.PauliZ(0))
+
+        tape.trainable_params = {1, 2, 3}
+
+        dev = qml.device("default.qubit", wires=1)
+        grad_A = tape.jacobian(dev, method="analytic")
+        grad_F = tape.jacobian(dev, method="numeric")
+
+        # gradient has the correct shape and every element is nonzero
+        assert grad_A.shape == (1, 3)
+        assert np.count_nonzero(grad_A) == 3
+        # the different methods agree
+        assert np.allclose(grad_A, grad_F, atol=tol, rtol=0)
+
+    def test_gradient_repeated_gate_parameters(self, mocker, tol):
+        """Tests that repeated use of a free parameter in a
+        multi-parameter gate yield correct gradients."""
+        dev = qml.device("default.qubit", wires=1)
+        params = np.array([0.8, 1.3], requires_grad=True)
+
+        def cost(params, method):
+            with AutogradInterface.apply(ReversibleTape()) as tape:
+                qml.RX(np.array(np.pi / 4, requires_grad=False), wires=[0])
+                qml.Rot(params[1], params[0], 2 * params[0], wires=[0])
+                expval(qml.PauliX(0))
+
+            tape.jacobian_options = {"method": method}
+            return tape.execute(dev)
+
+        spy_numeric = mocker.spy(ReversibleTape, "numeric_pd")
+        spy_analytic = mocker.spy(ReversibleTape, "analytic_pd")
+
+        grad_fn = qml.grad(cost)
+        grad_F = grad_fn(params, method="numeric")
+
+        spy_numeric.assert_called()
+        spy_analytic.assert_not_called()
+
+        grad_A = grad_fn(params, method="analytic")
+
+        spy_analytic.assert_called()
+
+        # the different methods agree
+        assert np.allclose(grad_A, grad_F, atol=tol, rtol=0)
+
+    # def test_gradient_parameters_inside_array(self, tol):

```

Comment:
This is the first test I have come across that shows behaviour supported by the old `Variable` system, and not the new tape system.

Previously, the `Variable` class would relentlessly flatten _everything_, even array parameters. Every element in the variable list was independently differentiable; e.g.,

```
qml.RX(a), qml.RY(b), qml.Hermitian(np.diag(c, d))  --> [a, b, c, 0, 0, d]
```

This allowed us to vary the value of parameter `c` independently, e.g., for finite difference evaluation.

With the tape, however, gate arguments are left unprocessed:
```
qml.RX(a), qml.RY(b), qml.Hermitian(np.diag(c, d))  --> [a, b, np.array([[c, 0], [0, d]])]
```

The only way to vary `c` independently is via a new quantum operation, `qml.DiagonalHermitian(c, d, wires=0)`:
```
qml.RX(a), qml.RY(b), qml.DiagonalHermitian(c, d)  --> [a, b, c, d]
```
To be honest, I don't think this is a bad thing---we have tried to keep all differentiable operations so far only accepting real parameters. I was actually pretty surprised that this edge case _worked_ in the old system, I have never seen it before porting this test.

Reply:
The (old) tests were a bit of a mess.

* Tests for `QubitQNode` spent most of the time testing different inputs to the qnode. E.g., float input, array input, fan out, keyword arguments, etc. All of these tests are redundant, so I hardly had any to port over. I wrote most of the new tests by hand, and added in the parmeter-shift tests from `test_quantum_gradient.py`.

* Tests for the `CVNode` also spent most of the time testing different inputs to the qnode. I wrote most of the new tests by hand, and added in the CV tests from `test_quantum_gradient.py`.

* For this PR, I ported over all the tests from `test_qnode_reversible.py`. However, like above, almost all tests were concerned with qnode input style. I tried to port over as many as I could.